Firestore/core/test/firebase/firestore/nanopb/nanopb_testing.h

```diff
@@ -35,21 +37,27 @@ namespace nanopb {
  * in a blob.
  */
 template <typename T>
-nanopb::ByteString ProtobufSerialize(const T& protobuf_message) {
-  std::vector<uint8_t> buffer(protobuf_message.ByteSizeLong());
-  bool ok = protobuf_message.SerializeToArray(buffer.data(),
-                                              static_cast<int>(buffer.size()));
+ByteString ProtobufSerialize(const T& protobuf_message) {
+  ByteStringWriter writer;
+
+  size_t size = protobuf_message.ByteSizeLong();
+  writer.Reserve(size);
+
+  bool ok =
+      protobuf_message.SerializeToArray(writer.pos(), static_cast<int>(size));
 
   // SerializeToArray can only fail if the buffer wasn't large enough.
   HARD_ASSERT(ok);
-  return nanopb::ByteString(buffer);
+
+  writer.SetSize(size);
```

Comment:
This is a little error-prone, though I understand it's hard to avoid. One random idea is something like
```cpp
bool ok = writer.WriteDirectly(size, [&](uint8_t* pos, size_t size) {
  return protobuf_message.SerializeToArray(pos, static_cast<int>(size));
});

template <typename ReturnT, typename CallbackT>
auto ByteStringWriter::WriteDirectly(size_t size, CallbackT callback) -> decltype(callback()) {
  Reserve(size);
  auto result = callback(pos(), size);
  SetSize(size);
  return result;
}
```
Not necessarily worth the trouble, just throwing it out here. Maybe `pos` and `SetSize` (and even `Reserve`) could become private in that case.

Reply:
Hmm, I've only noticed the usage in this file, where `SerializeToArray` appears to return a `bool`, have I missed something? However, I agree that at this point it's probably more trouble than it's worth; we can revisit later on if this functionality sees more usage.

Reply:
SerializeToArray returns a bool but then we immediately assert success, because the only way we could have failed was to pass a buffer smaller than `message.ByteSizeLong()`. In my initial attempt of this, I just immediately made that assertion inside the lambda and returned nothing.

If we implement anything else that uses this API we should revisit, and how that interaction fits together might guide our redesign. For now I'm going to leave it.